unbalanced_dataset/under_sampling/instance_hardness_threshold.py

```diff
@@ -0,0 +1,240 @@
+            the majority class.
+
+        random_state : int or None, optional (default=None)
+            Seed for random number generation.
+
+        verbose : bool, optional (default=True)
+            Boolean to either or not print information about the processing
+
+        n_jobs : int, optional (default=-1)
+            The number of thread to open when it is possible.
+
+        Returns
+        -------
+        None
+
+        """
+        super(InstanceHardnessThreshold, self).__init__(
+            return_indices=return_indices,
+            random_state=random_state,
+            verbose=verbose)
+
+        self.estimator = estimator
+        self.threshold = threshold
+
+        possible_modes = ('maj', 'all')
+        if mode  not in possible_modes:
+            raise ValueError('Unknown mode parameter.')
+        else:
+            self.mode = mode
+
+        self.cv = cv
+
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Find the classes statistics before to perform sampling.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Matrix containing the data which have to be sampled.
+
+        y : ndarray, shape (n_samples, )
+            Corresponding label for each sample in X.
+
+        Returns
+        -------
+        self : object,
+            Return self.
+
+        """
+        # Check the consistency of X and y
+        X, y = check_X_y(X, y)
+
+        super(InstanceHardnessThreshold, self).fit(X, y)
+
+        return self
+
+    def transform(self, X, y):
+        """Resample the dataset.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Matrix containing the data which have to be sampled.
+
+        y : ndarray, shape (n_samples, )
+            Corresponding label for each sample in X.
+
+        Returns
+        -------
+        X_resampled : ndarray, shape (n_samples_new, n_features)
+            The array containing the resampled data.
+
+        y_resampled : ndarray, shape (n_samples_new)
+            The corresponding label of `X_resampled`
+
+        idx_under : ndarray, shape (n_samples, )
+            If `return_indices` is `True`, a boolean array will be returned
+            containing the which samples have been selected.
+
+        """
+        # Check the consistency of X and y
+	X, y = check_X_y(X, y)
+
+        super(InstanceHardnessThreshold, self).transform(X, y)
+
+
+        skf = StratifiedKFold(y, n_folds=self.cv, shuffle=False, 
+                random_state=self.rs_)
+
+
+        probabilities = np.zeros(y.shape[0], dtype=float)
+
+        for train_index, test_index in skf:
+            X_train, X_test = X[train_index], X[test_index]
+            y_train, y_test = y[train_index], y[test_index]
+
+            self.estimator.fit(X_train, y_train)
+
+            probs = self.estimator.predict_proba(X_test)
```

Comment:
done.

143         if not hasattr(estimator, 'predict_proba'):
144             raise ValueError('Estimator does not have predict_proba method.')
145         else:
146             self.estimator = estimator